full-service/src/json_rpc/wallet_api.rs

```diff
@@ -0,0 +1,877 @@
+            .header(ContentType::JSON)
+            .body(body.to_string())
+            .dispatch();
+        assert_eq!(res.status(), Status::Ok);
+        let body = res.body().unwrap().into_string().unwrap();
+        // log::info!(logger, "Attempted dispatch got response {:?}", body);
+        let res: JsonValue = serde_json::from_str(&body).unwrap();
+        res.get("result").unwrap().clone()
+    }
+
+    // fn dispatch_expect_error(
+    //     client: &Client,
+    //     body: JsonValue,
+    //     logger: &Logger,
+    //     expected_err: String,
+    // ) {
+    //     let mut res = client
+    //         .post("/wallet")
+    //         .header(ContentType::JSON)
+    //         .body(body.to_string())
+    //         .dispatch();
+    //     assert_eq!(res.status(), Status::Ok);
+    //     let body = res.body().unwrap().into_string().unwrap();
+    //     log::info!(logger, "Attempted dispatch got response {:?}", body);
+    //     assert_eq!(body, expected_err);
+    // }
+
+    #[test_with_logger]
+    // A create_account RPC should correctly invoke the wallet service's create_account method.
+    fn test_create_account(logger: Logger) {
+        let rocket_config = rocket::Config::build(rocket::config::Environment::Development)
+            .port(get_free_port())
+            .unwrap();
+
+        // Create Account request body.
+        let body = json!({
+            "method": "create_account",
+            "params": {
+                "name": "Alice Main Account",
+            }
+        });
+
+        let expected_response: json_rpc::CreateAccountResponse = {
+            let mut account = json_rpc::Account::default();
+            account.account_id = "123".to_string();
+            json_rpc::CreateAccountResponse {
+                entropy: "ABC".to_string(),
+                account,
+            }
+        };
+
+        let mut mock_wallet_service = MockWallet::new();
+        {
+            let expected = expected_response.clone();
+            // create_account should be called exactly once.
+            mock_wallet_service
+                .expect_create_account()
+                .return_once(move |_name, _first_block| Ok(expected));
+        }
+
+        let state = WalletApiState {
+            service: Arc::new(Mutex::new(mock_wallet_service)),
+        };
+        let rocket = rocket(rocket_config, state);
+        let client = Client::new(rocket).unwrap();
+
+        // The result should be the expected response.
+        let result = dispatch(&client, body, &logger);
+        assert!(result.get("entropy").is_some());
+        assert_eq!(
+            result.get("entropy").unwrap(),
+            expected_response.entropy.as_str()
+        );
+
+        let account_obj = result.get("account").unwrap();
+        assert_eq!(
+            account_obj.get("account_id").unwrap(),
+            expected_response.account.account_id.as_str()
+        );
+    }
+
+    // #[test_with_logger]
```

Comment:
Is this intentional?

Reply:
Basically, yes. This PR introduces the ability to mock the wallet service, and so the unit tests in this file should pretty much just test that the "plumbing" from JSON, to Rust, into the mock wallet service, and back to the JSON response. The commented-out unit tests below are more like "end-to-end" tests because they actually set up a database and test the wallet service. I'd like to move this kind of testing to the wallet service itself (so that json_rpc isn't involved in testing that service), and / or to crate-level integration tests

Reply:
However, maybe it's best to restore these tests now, and deal with them in a follow-up PR? This PR has already gotten too big

Reply:
I disagree with this - these tests were designed to be end-to-end tests, and I would prefer if they stay. We can add additional tests at lower layers, but imo it's important to test the actual rpc endpoints that the API is delivering.